zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -808,74 +840,163 @@ public int compare(PathWithStat left, PathWithStat right) {
+                 index < targetChildren.size() && paginatedChildren.size() < maxReturned;
+                 index++) {
+                String child = targetChildren.get(index).getPath();
+                packetLength += child.length() + PAGINATION_PACKET_CHILD_EXTRA_BYTES;
+                if (!canPacketFitInMaxBuffer(packetLength)) {
+                    // Stop adding more children to ensure packet is below max buffer
+                    break;
+                }
+                paginatedChildren.add(child);
             }
 
-            // This is the last page, set the watch
-            if (childrenQueue.isEmpty()) {
+            // Decrement index as it was incremented once before exiting the for-loop.
+            index--;
+
+            if (index == targetChildren.size() - 1 && watcher != null) {
+                // All children are added so this is the last page, set the watch
                 childWatches.addWatch(path, watcher);
             }
-            return result;
         }
+
+        updateNextPage(nextPage, targetChildren, index);
+        updateReadStat(path, countReadChildrenBytes(paginatedChildren));
+
+        return paginatedChildren;
+    }
+
+    private boolean canPacketFitInMaxBuffer(int packetLength) {
+        return packetLength <= BinaryInputArchive.maxBuffer;
+    }
+
+    private int countReadChildrenBytes(Collection<String> children) {
+        return children.stream().mapToInt(String::length).sum();
+    }
+
+    private void buildChildrenPathWithStat(DataNode n, String path, Stat stat, long minCzxId,
+                                           List<PathWithStat> targetChildren) {
+        synchronized (n) {
+            if (stat != null) {
+                n.copyStat(stat);
+            }
+            for (String child : n.getChildren()) {
+                DataNode childNode = nodes.get(path + "/" + child);
+                if (null != childNode) {
+                    final long czxId = childNode.stat.getCzxid();
+
+                    if (czxId < minCzxId) {
+                        // Filter out nodes that are below minCzxId
+                        continue;
+                    }
+
+                    Stat childStat = new Stat();
+                    childNode.copyStat(childStat);
+
+                    // Cannot discard before having sorted and removed offset
+                    targetChildren.add(new PathWithStat(child, childStat));
+                }
+            }
+        }
+    }
+
+    /*
+     * totalLength = (PAGINATION_PACKET_CHILD_EXTRA_BYTES + childLength) * numChildren + PAGINATION_PACKET_BASE_BYTES
+     */
+    private int computeChildrenPacketLength(Collection<String> children) {
+        int length = children.stream().mapToInt(child -> PAGINATION_PACKET_CHILD_EXTRA_BYTES + child.length()).sum();
+        return length + PAGINATION_PACKET_BASE_BYTES;
+    }
+
+    private void updateNextPage(PaginationNextPage nextPage, List<PathWithStat> children, int lastAddedIndex) {
```

Comment:
The log seems to be returning the previously fetched item xid and offset. Shouldn't we return the "NEXT" one?

Reply:
czxid is the last one in the previous page, offset is the next one (last offset + 1). If the next offset doesn't exist in the next call, it'll just automatically get the next czxid and offset 0.

Reply:
@jiajunwang @dasahcc Updated the nextPage to have the exact minCzxid and offset.
Eg. same czxid 100, the first 10 children are returned, so next page is with starting offset 10: {100, 10},
If czxid is uniq, next page is with offset 0: {101, 0}